src/python/pants/core/target_types_test.py

```diff
@@ -128,3 +138,108 @@ def assert_prefix_mapping(
+    * An `archive` containing another `archive`.
+    """
+
+    rule_runner = RuleRunner(
+        rules=[
+            *target_type_rules(),
+            *pex_from_targets.rules(),
+            *package_python_binary.rules(),
+            QueryRule(BuiltPackage, [ArchiveFieldSet]),
+        ],
+        target_types=[ArchiveTarget, Files, RelocatedFiles, PythonBinary],
+    )
+    rule_runner.set_options(
+        ["--backend-packages=pants.backend.python", "--no-pants-distdir-legacy-paths"]
+    )
+
+    rule_runner.create_file("resources/d1.json", "{'k': 1}")
+    rule_runner.create_file("resources/d2.json", "{'k': 2}")
+    rule_runner.add_to_build_file(
+        "resources",
+        dedent(
+            """\
+            files(name='original_files', sources=['*.json'])
+
+            relocated_files(
+                name='relocated_files',
+                files_targets=[':original_files'],
+                src="resources",
+                dest="data",
+            )
+            """
+        ),
+    )
+
+    rule_runner.create_file("project/app.py", "print('hello world!')")
+    rule_runner.add_to_build_file("project", "python_binary(sources=['app.py'])")
+
+    rule_runner.add_to_build_file(
+        "",
+        dedent(
+            """\
+            archive(
+                name="archive1",
+                packages=["project"],
+                files=["resources:original_files"],
+                format="zip",
+            )
+
+            archive(
+                name="archive2",
+                packages=[":archive1"],
+                files=["resources:relocated_files"],
+                format="tar",
+            )
+            """
+        ),
+    )
+
+    def get_archive(target_name: str) -> FileContent:
+        tgt = rule_runner.get_target(Address("", target_name=target_name))
+        built_package = rule_runner.request(BuiltPackage, [ArchiveFieldSet.create(tgt)])
+        digest_contents = rule_runner.request(DigestContents, [built_package.digest])
+        assert len(digest_contents) == 1
+        return digest_contents[0]
+
+    def assert_archive1_is_valid(zip_bytes: bytes) -> None:
+        io = BytesIO()
+        io.write(zip_bytes)
+        with zipfile.ZipFile(io) as zf:
+            assert set(zf.namelist()) == {
+                "resources/d1.json",
+                "resources/d2.json",
+                "project/project.pex",
+            }
+            with zf.open("resources/d1.json", "r") as f:
+                assert f.read() == b"{'k': 1}"
+            with zf.open("resources/d2.json", "r") as f:
+                assert f.read() == b"{'k': 2}"
+
+    archive1 = get_archive("archive1")
+    assert_archive1_is_valid(archive1.content)
+
+    archive2 = get_archive("archive2")
+    # For some reason, writing the result to BytesIO, then passing via `fileobj` to
+    # `tarfile.open()` does not work properly, even if calling `.flush()` before. So we write to
+    # disk.
```

Comment:
I struggled with this for way too long to try to get it working.

I checked that things work correctly when running `./pants package` and then externally calling `tar xf`. And I think this test is still comprehensive, only a little clunky.

Reply:
Did you reset the position to the beginning? https://stackoverflow.com/questions/53485708/how-the-write-read-and-getvalue-methods-of-python-io-bytesio-work

Reply:
That would do it...thank you!